app/Wrappers/SecretStreamWrapper.php

```diff
@@ -0,0 +1,187 @@
+     *
+     * @param $data the plaintext to write
+     * @return int how many bytes of input we used, else 0 if everything went wrong
+     */
+    public function stream_write($data) {
+        // Append data to our internal buffer.
+        $this->buffer .=  $data;
+
+        // Write out as many full-sized messages as we can without fully depleting the buffer.
+        if ($this->flushBuffer()) {
+            // We wrote everything provided to us.
+            return strlen($data);
+        }
+
+        // We failed to write.
+        return 0;
+    }
+
+    /**
+     * Flush out our final data and close all resources. e.g. `fclose()` would call this.
+     */
+    public function stream_close() {
+        // Write out data as encrypted messages until we have only the final message's data in the buffer.
+        $this->flushBuffer();
+
+        // Flush out the final message, tagged as such.
+        if (strlen($this->buffer) > 0) {
+            $encrypted = sodium_crypto_secretstream_xchacha20poly1305_push(
+                $this->stream,
+                $this->buffer,
+                '', // This is an inactive placeholder.
+                SODIUM_CRYPTO_SECRETSTREAM_XCHACHA20POLY1305_TAG_FINAL
+            );
+            fwrite($this->wrapped, $encrypted); // TODO : Can an error here be caught somehow?
+        }
+
+        // Close our wrapped resource.
+        fclose($this->wrapped);
+    }
+
+    /**
+     * This is used to flush out our internal buffer, clearing everything until we only have enough left to send a final
+     * message, which we have to tag (and so send in stream_close()).
+     *
+     * @return bool whether all writes succeeded
+     */
+    private function flushBuffer() {
+        // While we can send a full-sized message and still have data left for the final message...
+        while (strlen($this->buffer) > self::MESSAGE_SIZE) {
+            // Write out a message's worth of data from the start of the buffer.
+            $encrypted = sodium_crypto_secretstream_xchacha20poly1305_push(
+                $this->stream,
+                substr($this->buffer, 0, self::MESSAGE_SIZE)
+            );
+            $this->buffer = substr($this->buffer, self::MESSAGE_SIZE);
+            if (!fwrite($this->wrapped, $encrypted)) return false;
+        }
+        return true;
+    }
+
+    /**
+     * Get the key that we will use for encryption, as defined in the config (and so probably by the environment
+     * variable APP_KEY.)
+     *
+     * @return string the key itself to use
+     */
+    public static function getKey() {
+        $key = config('app.key');
+
+        // Sometimes, the key is encoded in base64. Decrypt it if this is the case.
+        if (Str::startsWith($key, 'base64:')) {
+            $key = base64_decode(substr($key, 7));
+        }
+
+        // Check that the key is the write length.
+        // TODO : Can we guarantee that app.key will be configured as 32 random bytes?
```

Comment:
If people use `artisan key:generate` then yes, it should be.

Reply:
I don't see much of a risk of anything insecure but also the right length being used here. We'll get a nice error spat out if the length is wrong too, so I think we should be able to leave this as is :+1: